src/plugins/intel_cpu/src/node.cpp

```diff
@@ -771,46 +771,61 @@ void Node::initDescriptor(const NodeConfig& config) {
     selectedPD->setConfig(updatedConfig);
 }
 
-void Node::prepareMemory(const std::vector<DnnlMemoryDescPtr>& intDescs) {
-    if (internalBlobs.size() != intDescs.size()) {
-        IE_THROW() << "Can't prepare memory for internal blob, internal blob and internal descs number do not match "
-                   << internalBlobs.size() << " vs " << intDescs.size();
+void Node::prepareMemory(const DnnlMemoryDescPtr& intDesc, size_t indx) {
+    size_t minSize = indx + 1;
+    if (internalBlobMemory.size() < minSize) {
+        internalBlobMemory.resize(minSize);
     }
 
-    internalBlobMemory.clear();
-    for (size_t i = 0; i < internalBlobs.size(); i++) {
-        const auto &internalBlob = internalBlobs[i];
+    if (minSize > internalBlobs.size()) {
+        IE_THROW() << "Can't prepare memory for internal blob, requested index: " << indx <<
+            " is out of bounds of the internalBlobs vector of size " << internalBlobs.size();
+    }
 
-        auto create = [&] () {
-            // TODO [DS]: internal blobs should be removed or rewritten using Memory object
-            auto newDesc = MemoryDescUtils::convertToDnnlBlockedMemoryDesc(internalBlob->getTensorDesc());
+    const auto &internalBlob = internalBlobs[indx];
 
-            Memory memory{ engine };
-            memory.Create(newDesc, internalBlob->buffer());
+    auto create = [&] () {
+        // TODO [DS]: internal blobs should be removed or rewritten using Memory object
+        auto newDesc = MemoryDescUtils::convertToDnnlBlockedMemoryDesc(internalBlob->getTensorDesc());
 
-            MemoryPtr _ptr = std::make_shared<Memory>(engine);
-            _ptr->Create(*intDescs[i]);
-            _ptr->SetData(memory);
+        Memory memory{ engine };
+        memory.Create(newDesc, internalBlob->buffer());
 
-            return _ptr;
-        };
+        MemoryPtr _ptr = std::make_shared<Memory>(engine);
+        _ptr->Create(intDesc);
+        node::Reorder::reorderData(memory, *_ptr, context->getParamsCache());
+        return _ptr;
+    };
 
-        MemoryPtr ptr;
-        auto weightCache = context->getWeightsCache();
-        if (weightCache != nullptr) {
-            const uint64_t data_hash = weightCache->GetHashFunc().hash(
-                    internalBlob->buffer(), internalBlob->byteSize());
+    MemoryPtr ptr;
+    auto weightCache = context->getWeightsCache();
+    if (weightCache != nullptr && memory::format_kind::blocked == intDesc->getDnnlDesc().get_format_kind()) {
```

Comment:
Except blocked and opaque format, could there be other format that we can cache the weights?

Reply:
At the moment oneDNN has four memory desc types: blocked, opaque, winograd, sparce. winograd format is used for winograd convolutions only. Sparce format at the moment is used for convolutions and fully connected nodes. But the `Node::prepareMemory(const std::vector<DnnlMemoryDescPtr>& intDescs)` call is used only for deconvolutions and rnn since they are two node types that still use internal blobs. Therefore, in this method it is enough to cache blocked format tensors only. Obviously, we have to revise the weight cache usage considering the dynamic shapes paradigm to avoid such condition checks.